packages/preview2-shim/lib/io/worker-thread.js

```diff
@@ -0,0 +1,312 @@
+        // if it blocks the main thread, we can't actually blocking write and flush
+        // instead we eagerly return. This is primarily for stdio. Note Node.js actually
+        // provides no blocking write to stdout mechanism on Windows.
+        // This can be resolved when components themselves move into workers
+        if (blocksMainThread) {
+          stream.write(payload);
+          return;
+        }
+        return new Promise((resolve, reject) => {
+          stream.write(payload, err => err
+              ? reject(streamError(id, stream, err))
+              : resolve(BigInt(payload.byteLength))
+          );
+        });
+      case calls.OUTPUT_STREAM_FLUSH:
+        var stream = getStreamOrThrow(id);
+        if (stream.flushPromise) return stream.flushPromise;
+        return stream.flushPromise = new Promise((resolve, reject) => {
+          stream.stream.write(new Uint8Array([]), err => err ? reject(streamError(id, stream, err)) : resolve());
+        }).then(() => {
+          stream.stream.flushPromise = null;
+        }, err => {
+          stream.stream.flushPromise = null;
+          throw streamError(id, stream.stream, err);
+        });
+      case calls.OUTPUT_STREAM_BLOCKING_FLUSH:
+        var { stream } = getStreamOrThrow(id);
+        return new Promise((resolve, reject) => {
+          stream.write(new Uint8Array([]), err => err ? reject(streamError(id, stream, err)) : resolve());
+        });
+      case calls.OUTPUT_STREAM_WRITE_ZEROES:
+        var { stream } = getStreamOrThrow(id);
+        return void stream.write(new Uint8Array(Number(payload)));
+      case calls.OUTPUT_STREAM_BLOCKING_WRITE_ZEROES_AND_FLUSH:
+        var { stream } = getStreamOrThrow(id);
+        return new Promise((resolve, reject) => {
+          stream.write(new Uint8Array(Number(payload)), err => err
+            ? reject(streamError(id, stream, err))
+            : resolve(BigInt(payload.byteLength))
+          )
+        });
+      case calls.OUTPUT_STREAM_SPLICE:
+        var { stream: outputStream } = getStreamOrThrow(id);
+        var { stream: inputStream } = getStreamOrThrow(payload.src);
+        var bytesRemaining = Number(payload.len);
+        var chunk;
+        while (bytesRemaining > 0 && (chunk = inputStream.read(Math.min(outputStream.writableHighWaterMark - stream.writableLength, bytesRemaining)))) {
+          bytesRemaining -= chunk.byteLength;
+          outputStream.write(chunk);
+        }
+        // TODO: these error handlers should be attached, and only for the duration of the splice flush
+        if (inputStream.errored) {
+          streamError(payload.src, inputStream, inputStream.errored); // error ignored?
+          throw streamError(id, outputStream, inputStream.errored);
+        }
+        if (outputStream.errored) {
+          // input stream not closed?
+          throw streamError(id, outputStream, outputStream.errored);
+        }
+        return payload.len - BigInt(bytesRemaining);
```

Comment:
is `payload.len` always a positive integer? If it's 0 then this result would be negative. Is that expected?

Reply:
In the component model itself, this is typed as a u64, so it's not possible to be negative.

For components that call to WASI directly, this is therefore always guaranteed.

Reply:
Secondly, because `bytesRemaining` is originally set to `payload.len`, `bytesRemaining` can never be > 0 when `payload.len` is `0` here.

Reply:
Got it. Thanks.